File: source/go/manual_linked_list.go
package main

import "fmt"

// Manual linked list simulated with Go structs (equivalent to alloc/free pattern)
type Node struct {
	val  int
	next *Node
}

var head *Node

func push(val int) {
	head = &Node{val: val, next: head}
}

func traverse() {
	fmt.Print("list:")
	current := head
	for current != nil {
		fmt.Printf(" %d", current.val)
		current = current.next
	}
	fmt.Println()
}

func freeAll() {
	count := 0
	current := head
	for current != nil {
		next := current.next
		current = nil // allow GC
		count++
		current = next
	}
	head = nil
	fmt.Printf("freed %d nodes\n", count)
}

func main() {
	push(10)
	push(20)
	push(30)
	traverse()
	freeAll()
}

File: source/javascript/manual_linked_list.js
// Manual linked list simulation using JS objects
// (simulating the Hemlock alloc/ptr approach with a simple node model)
let head = null;

function push(val) {
    head = { val, next: head };
}

function traverse() {
    const parts = [];
    let current = head;
    while (current !== null) { parts.push(current.val); current = current.next; }
    process.stdout.write("list:");
    for (const v of parts) process.stdout.write(` ${v}`);
    process.stdout.write("\n");
}

function freeAll() {
    let current = head;
    let count = 0;
    while (current !== null) { const next = current.next; count++; current = next; }
    head = null;
    console.log(`freed ${count} nodes`);
}

push(10); push(20); push(30);
traverse();
freeAll();
